### PR TITLE
Persistence

### PR DESCRIPTION
**Pull request by HandlessMidas**

### CODE REVIEW

File: storage/postgresStorage.go [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+	row := tx.QueryRow("select g.last_user_id from GeneralInfo g where g.base_id = 0")
+
+	lastId := 0
+	_ = row.Scan(&lastId)
+	id := strconv.Itoa(lastId)
+	lastId += 1
+
+	_, _ = tx.ExecContext(ctx, "update GeneralInfo set last_user_id = $1 where base_id = 0", lastId)
+
+	_ = tx.Commit()
```
Comment (M0nsteR) [inline storage/postgresStorage.go:38]: Do you see any concurrency risks here?

File: initdb.sql [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+create table Users(
+    id int primary key,
+    login text
+);
+
+create table Password(
+    id int primary key,
+    password bytea,
+    constraint fr_user_id foreign key(id) references Users(id)
```
Comment (M0nsteR) [inline initdb.sql:18]: Do you really need this field to be primary and foreign key at the same time? Do you really need this Password table at all? Please explain your reasoning behind this decision.

Comment (ilyavoronin) [reply initdb.sql:18]: The "id" field could be just unique. And it should be a foreign key to correspond to id from Users table. 
Password table could be moved to another database later for security reasons. So this is the main reason to have password and login in separate tables

Comment (M0nsteR) [reply initdb.sql:18]: I can see no extra security benefits here, just transactional and computational penalties. BTW, have you ever heard of a salted hash?

File: storage/postgresStorage.go (hunk 1 above)
Comment (M0nsteR) [reply storage/postgresStorage.go:38]: What happens when two concurrent transactions assign the same user id?

File: api/api.go [hunk 3]
```diff
@@ -54,14 +52,19 @@ func (a *Api) register(w http.ResponseWriter, r *http.Request) {
...
 	if err != nil {
 		w.WriteHeader(http.StatusBadRequest)
 		if err == model.ErrAlreadyExists {
 			_, _ = w.Write([]byte("login already exists"))
 		}
 		return
 	}
+	respJson, err := json.Marshal(user)
+	_, err = w.Write(respJson)
+	if err != nil {
```
Comment (NekoNe) [inline api/api.go:65]: Could you explain what is the point of this check, please?

File: storage/postgresStorage.go (hunk 1 above)
Comment (HandlessMidas) [reply storage/postgresStorage.go:38]: Added mutexes to fix concurrency

File: api/api.go (hunk 3 above)
Comment (HandlessMidas) [reply api/api.go:65]: Now it writes an error status to the header

File: initdb.sql (hunk 2 above)
Comment (ilyavoronin) [reply initdb.sql:18]: I think GenerateFromPassword method from bcrypt lib uses salt internally

Comment (M0nsteR) [reply initdb.sql:18]: That's right. So your hash strings are already made hard to break. The salt is there to require each password to be attacked individually. You add no extra layer of security by merely placing these strings to a separate table or database, but the risks of data inconsistency are high. Just think about it.

**Review by M0nsteR — APPROVED** [review]
Let's move on, folks..

**Pull request merged**